File: backend/internal/protocol/decoration.go
package protocol

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

// ApplyDecoration evaluates a template like "{v/10}.{v%10}" using integer v.
func ApplyDecoration(template string, v int64) (string, error) {
	if template == "" {
		return "", nil
	}
	var out strings.Builder
	i := 0
	for i < len(template) {
		if template[i] == '{' {
			end := strings.IndexByte(template[i:], '}')
			if end < 0 {
				return "", fmt.Errorf("unclosed expression in decoration")
			}
			expr := strings.TrimSpace(template[i+1 : i+end])
			n, err := evalDecorationExpr(expr, v)
			if err != nil {
				return "", fmt.Errorf("decoration expr %q: %w", expr, err)
			}
			out.WriteString(strconv.FormatInt(n, 10))
			i += end + 1
		} else {
			out.WriteByte(template[i])
			i++
		}
	}
	return out.String(), nil
}

func fieldValueToInt(v interface{}) (int64, bool) {
	switch n := v.(type) {
	case uint64:
		return int64(n), true
	case int64:
		return n, true
	case int:
		return int64(n), true
	case float64:
		return int64(n), true
	default:
		return 0, false
	}
}

func applyFieldDecoration(result map[string]interface{}, name, decoration string, val interface{}) {
	if decoration == "" {
		return
	}
	iv, ok := fieldValueToInt(val)
	if !ok {
		return
	}
	dec, err := ApplyDecoration(decoration, iv)
	if err != nil {
		return
	}
	result[name+"_display"] = dec
}

type exprParser struct {
	src string
	pos int
	v   int64
}

func evalDecorationExpr(expr string, v int64) (int64, error) {
	p := &exprParser{src: strings.TrimSpace(expr), v: v}
	if p.src == "" {
		return 0, fmt.Errorf("empty expression")
	}
	n, err := p.parseExpr()
	if err != nil {
		return 0, err
	}
	p.skipSpace()
	if p.pos < len(p.src) {
		return 0, fmt.Errorf("unexpected trailing input at %q", p.src[p.pos:])
	}
	return n, nil
}

func (p *exprParser) parseExpr() (int64, error) {
	left, err := p.parseTerm()
	if err != nil {
		return 0, err
	}
	for {
		p.skipSpace()
		if p.pos >= len(p.src) {
			break
		}
		op := p.src[p.pos]
		if op != '+' && op != '-' {
			break
		}
		p.pos++
		right, err := p.parseTerm()
		if err != nil {
			return 0, err
		}
		if op == '+' {
			left += right
		} else {
			left -= right
		}
	}
	return left, nil
}

func (p *exprParser) parseTerm() (int64, error) {
	left, err := p.parseFactor()
	if err != nil {
		return 0, err
	}
	for {
		p.skipSpace()
		if p.pos >= len(p.src) {
			break
		}
		op := p.src[p.pos]
		if op != '*' && op != '/' && op != '%' {
			break
		}
		p.pos++
		right, err := p.parseFactor()
		if err != nil {
			return 0, err
		}
		switch op {
		case '*':
			left *= right
		case '/':
			if right == 0 {
				return 0, fmt.Errorf("division by zero")
			}
			left /= right
		case '%':
			if right == 0 {
				return 0, fmt.Errorf("modulo by zero")
			}
			left %= right
		}
	}
	return left, nil
}

func (p *exprParser) parseFactor() (int64, error) {
	p.skipSpace()
	if p.pos >= len(p.src) {
		return 0, fmt.Errorf("unexpected end of expression")
	}
	switch p.src[p.pos] {
	case '(':
		p.pos++
		n, err := p.parseExpr()
		if err != nil {
			return 0, err
		}
		p.skipSpace()
		if p.pos >= len(p.src) || p.src[p.pos] != ')' {
			return 0, fmt.Errorf("missing closing parenthesis")
		}
		p.pos++
		return n, nil
	case 'v':
		p.pos++
		return p.v, nil
	default:
		start := p.pos
		if p.src[p.pos] == '-' {
			p.pos++
		}
		for p.pos < len(p.src) && unicode.IsDigit(rune(p.src[p.pos])) {
			p.pos++
		}
		if start == p.pos || (p.src[start] == '-' && p.pos == start+1) {
			return 0, fmt.Errorf("invalid token at %q", p.src[p.pos:])
		}
		n, err := strconv.ParseInt(p.src[start:p.pos], 10, 64)
		if err != nil {
			return 0, fmt.Errorf("invalid number %q", p.src[start:p.pos])
		}
		return n, nil
	}
}

func (p *exprParser) skipSpace() {
	for p.pos < len(p.src) && unicode.IsSpace(rune(p.src[p.pos])) {
		p.pos++
	}
}

File: dashboard/src/lib/decoration.ts
/** Evaluate decoration template like "{v/10}.{v%10}" with integer v. */
export function applyDecoration(template: string, v: number): string {
  if (!template) return '';
  let out = '';
  let i = 0;
  while (i < template.length) {
    if (template[i] === '{') {
      const end = template.indexOf('}', i);
      if (end < 0) throw new Error('unclosed expression in decoration');
      const expr = template.slice(i + 1, end).trim();
      out += String(evalDecorationExpr(expr, v));
      i = end + 1;
    } else {
      out += template[i];
      i++;
    }
  }
  return out;
}

export function fieldValueToInt(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  if (typeof v === 'number' && Number.isFinite(v)) return Math.trunc(v);
  if (typeof v === 'bigint') return Number(v);
  return null;
}

export function applyFieldDecoration(_name: string, decoration: string | undefined, val: unknown): string | null {
  if (!decoration) return null;
  const iv = fieldValueToInt(val);
  if (iv === null) return null;
  try {
    return applyDecoration(decoration, iv);
  } catch {
    return null;
  }
}

function evalDecorationExpr(expr: string, v: number): number {
  const tokens = tokenize(expr.replace(/\bv\b/g, ` ${v} `));
  let pos = 0;

  const parseExpr = (): number => {
    let left = parseTerm();
    while (pos < tokens.length && (tokens[pos] === '+' || tokens[pos] === '-')) {
      const op = tokens[pos++];
      const right = parseTerm();
      left = op === '+' ? left + right : left - right;
    }
    return left;
  };

  const parseTerm = (): number => {
    let left = parseFactor();
    while (pos < tokens.length && (tokens[pos] === '*' || tokens[pos] === '/' || tokens[pos] === '%')) {
      const op = tokens[pos++];
      const right = parseFactor();
      if (op === '*') left *= right;
      else if (op === '/') left = Math.trunc(left / right);
      else left %= right;
    }
    return left;
  };

  const parseFactor = (): number => {
    const t = tokens[pos];
    if (t === '(') {
      pos++;
      const n = parseExpr();
      if (tokens[pos] !== ')') throw new Error('missing closing parenthesis');
      pos++;
      return n;
    }
    if (t === '-' && tokens[pos + 1] !== undefined) {
      pos++;
      return -parseFactor();
    }
    if (t === undefined || t === ')' || t === '+' || t === '*' || t === '/' || t === '%') {
      throw new Error('unexpected end of expression');
    }
    pos++;
    const n = Number(t);
    if (!Number.isFinite(n)) throw new Error(`invalid number ${t}`);
    return Math.trunc(n);
  };

  const result = parseExpr();
  if (pos < tokens.length) throw new Error('unexpected trailing input');
  return result;
}

function tokenize(expr: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  while (i < expr.length) {
    const c = expr[i];
    if (/\s/.test(c)) {
      i++;
      continue;
    }
    if ('()+-*/%'.includes(c)) {
      tokens.push(c);
      i++;
      continue;
    }
    if (/[\d-]/.test(c)) {
      let j = i;
      if (c === '-') j++;
      while (j < expr.length && /\d/.test(expr[j])) j++;
      if (j > i + (c === '-' ? 1 : 0)) {
        tokens.push(expr.slice(i, j));
        i = j;
        continue;
      }
    }
    throw new Error(`invalid token at ${expr.slice(i)}`);
  }
  return tokens;
}
